File: server-py/routes/calendar.py
"""ICS calendar parsing route."""
from fastapi import APIRouter, Request, HTTPException
from models import CalendarEventResponse, ParseIcsResponse
import re

router = APIRouter(prefix="/api/calendar", tags=["calendar"])


@router.post("/parse-ics")
async def parse_ics(req: Request):
    body = await req.json()
    ics = body.get("icsContent", "")
    if not ics or not isinstance(ics, str):
        raise HTTPException(400, detail={"error": "请提供 ICS 文件内容"})
    try:
        events = _parse_ics(ics)
        return {"success": True, "data": {"events": [e.model_dump() for e in events]}}
    except Exception as e:
        raise HTTPException(500, detail={"error": f"ICS 解析失败: {e}"})


def _parse_ics(content: str) -> list[CalendarEventResponse]:
    events: list[CalendarEventResponse] = []
    lines = content.split("\n")
    in_event = False
    cur: dict = {}

    for line in lines:
        line = line.strip()
        if line == "BEGIN:VEVENT":
            in_event = True
            cur = {"attendees": []}
        elif line == "END:VEVENT":
            if in_event and cur.get("id") and cur.get("title") and cur.get("startTime") and cur.get("endTime"):
                events.append(CalendarEventResponse(**cur))
            in_event = False
            cur = {}
        elif in_event:
            if ":" not in line:
                continue
            key, val = line.split(":", 1)
            base_key = key.split(";")[0]
            if base_key == "UID":
                cur["id"] = val
            elif base_key == "SUMMARY":
                cur["title"] = _decode_ics(val)
            elif base_key == "DTSTART":
                cur["startTime"] = _parse_ics_dt(val)
            elif base_key == "DTEND":
                cur["endTime"] = _parse_ics_dt(val)
            elif base_key == "LOCATION":
                cur["location"] = _decode_ics(val)
            elif base_key == "ORGANIZER":
                cur["organizer"] = _decode_ics(val.replace("mailto:", "", 1))
            elif base_key == "ATTENDEE":
                attendee = _decode_ics(val.replace("mailto:", "", 1))
                if attendee and attendee not in cur.get("attendees", []):
                    cur["attendees"].append(attendee)
    return events


def _parse_ics_dt(val: str) -> str:
    cleaned = val.replace("Z", "")
    if len(cleaned) == 8:
        return f"{cleaned[:4]}-{cleaned[4:6]}-{cleaned[6:8]}T00:00:00"
    if len(cleaned) == 15:
        return f"{cleaned[:4]}-{cleaned[4:6]}-{cleaned[6:8]}T{cleaned[9:11]}:{cleaned[11:13]}:{cleaned[13:15]}"
    return val


def _decode_ics(text: str) -> str:
    return text.replace("\\n", "\n").replace("\\,", ",").replace("\\;", ";").replace("\\\\", "\\")

File: server/routes/calendar.ts
import { Router } from "express";
import type { CalendarEventResponse, ParseIcsResponse, ApiError } from "@shared/api.interface";

const router = Router();

/**
 * POST /api/calendar/parse-ics
 * 解析上传的 ICS 日历文件，提取事件列表
 */
router.post("/parse-ics", async (req, res) => {
  try {
    const { icsContent } = req.body;

    if (!icsContent || typeof icsContent !== "string") {
      res.status(400).json({
        error: "缺少 ICS 文件内容",
        details: { icsContent: ["请提供 ICS 文件内容"] },
      } satisfies ApiError);
      return;
    }

    const events = parseIcsContent(icsContent);

    const response: ParseIcsResponse = {
      events,
    };

    res.json({ success: true, data: response });
  } catch (err: any) {
    console.error("ICS parse error:", err);
    res.status(500).json({
      error: "ICS 文件解析失败",
      details: { message: [err.message || "未知错误"] },
    } satisfies ApiError);
  }
});

/**
 * 解析 ICS 文件内容，提取 VEVENT 事件
 */
function parseIcsContent(content: string): CalendarEventResponse[] {
  const events: CalendarEventResponse[] = [];
  
  // 分割成 VEVENT 块
  const lines = content.split(/\r?\n/);
  let inEvent = false;
  let currentEvent: Partial<CalendarEventResponse> | null = null;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();

    if (line === "BEGIN:VEVENT") {
      inEvent = true;
      currentEvent = {
        attendees: [],
      };
    } else if (line === "END:VEVENT") {
      if (inEvent && currentEvent) {
        // 验证必要字段
        if (currentEvent.id && currentEvent.title && currentEvent.startTime && currentEvent.endTime) {
          events.push({
            id: currentEvent.id,
            title: currentEvent.title,
            startTime: currentEvent.startTime,
            endTime: currentEvent.endTime,
            attendees: currentEvent.attendees || [],
            location: currentEvent.location,
            organizer: currentEvent.organizer,
            source: "ics",
          });
        }
      }
      inEvent = false;
      currentEvent = null;
    } else if (inEvent && currentEvent) {
      // 解析事件属性
      const colonIndex = line.indexOf(":");
      if (colonIndex === -1) continue;

      const key = line.substring(0, colonIndex);
      const value = line.substring(colonIndex + 1);

      // 处理带参数的键（如 DTSTART;VALUE=DATE:20260115）
      const baseKey = key.split(";")[0];

      switch (baseKey) {
        case "UID":
          currentEvent.id = value;
          break;
        case "SUMMARY":
          currentEvent.title = decodeIcsText(value);
          break;
        case "DTSTART":
          currentEvent.startTime = parseIcsDateTime(value);
          break;
        case "DTEND":
          currentEvent.endTime = parseIcsDateTime(value);
          break;
        case "LOCATION":
          currentEvent.location = decodeIcsText(value);
          break;
        case "ORGANIZER":
          currentEvent.organizer = decodeIcsText(value.replace(/^mailto:/, ""));
          break;
        case "ATTENDEE":
          const attendee = decodeIcsText(value.replace(/^mailto:/, ""));
          if (attendee && !currentEvent.attendees!.includes(attendee)) {
            currentEvent.attendees!.push(attendee);
          }
          break;
      }
    }
  }

  return events;
}

/**
 * 解析 ICS 日期时间格式
 * 支持格式：20260115T090000Z, 20260115T090000, 20260115
 */
function parseIcsDateTime(value: string): string {
  // 移除时区标识
  const cleaned = value.replace(/Z$/, "");

  if (cleaned.length === 8) {
    // 日期格式：20260115
    const year = cleaned.substring(0, 4);
    const month = cleaned.substring(4, 6);
    const day = cleaned.substring(6, 8);
    return `${year}-${month}-${day}T00:00:00`;
  } else if (cleaned.length === 15) {
    // 日期时间格式：20260115T090000
    const year = cleaned.substring(0, 4);
    const month = cleaned.substring(4, 6);
    const day = cleaned.substring(6, 8);
    const hour = cleaned.substring(9, 11);
    const minute = cleaned.substring(11, 13);
    const second = cleaned.substring(13, 15);
    return `${year}-${month}-${day}T${hour}:${minute}:${second}`;
  }

  // 如果无法解析，返回原值
  return value;
}

/**
 * 解码 ICS 文本（处理转义字符）
 */
function decodeIcsText(text: string): string {
  return text
    .replace(/\\n/g, "\n")
    .replace(/\\,/g, ",")
    .replace(/\\;/g, ";")
    .replace(/\\\\/g, "\\");
}

export default router;
